### PR TITLE
Cache the slint! macro output token stream (rust-analyzer fast path)

### PR DESCRIPTION
**Pull request by tilladam**
## Motivation

Expanding the `slint!` macro recompiles the builtins, the selected style and any imported widgets on *every* invocation — tens of milliseconds even for a tiny snippet. rust-analyzer runs proc macros in a long-lived process and re-expands them very frequently (typically while the user edits *elsewhere*, so the macro body is unchanged), which makes this a hot path and a noticeable source of editor latency.

## Approach

Add a process-global cache of the macro's generated output, keyed by a hash of the macro body plus everything else that influences code generation (config fields, relevant env vars, source path) and validated against the **contents** of every loaded external file (the same set that drives the `include_bytes!` recompile markers). On a hit the whole expansion — parse, dependency loading, ~60 passes, codegen — is skipped and the cached output is re-parsed into a token stream.

The cached value is a `String`, not a `TokenStream`, because the cache is shared across the fresh thread rust-analyzer spawns per expansion, so entries must be `Send + Sync`. A `proc_macro::TokenStream` is neither (it is a handle into the compiler bridge, valid only within one expansion on one thread), so the portable cached form is the serialized source text, re-parsed on a hit.

It lives in `i-slint-compiler` next to the `CompilerConfiguration`/`Token` types its key is built from (gated behind the existing `rust` feature, `#[doc(hidden)]`) and is driven by the `slint-macros` proc macro.

- **On by default under rust-analyzer**, where the macro emits the compact live-preview output so re-parsing is cheap and re-expansions are frequent.
- **Off by default for ordinary `cargo build`** (the macro runs at most once per crate build, so populating a cache there is wasted work); opt in with `SLINT_MACRO_CACHE=1`.
- Bounded with FIFO eviction; dependency hashes are validated *after* releasing the lock so concurrent expansions don't serialize on disk reads.

## Impact (live-preview / rust-analyzer path, release, in-process medians)

| Path | Per expansion | vs base |
|---|---|---|
| No cache (base) | ~32 ms | — |
| Cache **miss** (compile + populate) | ~32.5 ms | ~noise |
| Cache **hit** (key + lookup + re-parse) | ~57 µs | **~560×** |

Amortized over re-expansions of an unchanged body (the rust-analyzer pattern): ~10× at 10 re-expansions, ~86× at 100. The miss path adds no meaningful overhead, so enabling it is effectively free.

## Tests & benchmarks

- Unit tests for the cache mechanics: hit/miss, dependency-change + file-deletion invalidation, FIFO eviction bound.
- An end-to-end replay test (two identical-body modules) that compiles a cached, re-parsed expansion.
- A `cache_simulation` benchmark module (`internal/compiler/benches/semantic_analysis.rs`) measuring miss vs hit for both codegen paths, the two cache-key strategies, and a head-to-head using the real `key_material`.

## Notes / limitations

- The cache key enumerates the config fields and env vars that reach the **live-preview** generator (the default-on path); it is complete there. The `SLINT_MACRO_CACHE` full-codegen opt-in is documented as requiring other codegen-affecting env (e.g. `SLINT_EMBED_RESOURCES`, `SLINT_SCALE_FACTOR`) to be held constant. A type-derived fingerprint to make this robust against future config fields is a sensible follow-up.

History is structured into two independently-buildable commits: the feature (all behavior change) and the benchmarks.

### CODE REVIEW

**Comment by ogoffart** [comment]
Thanks, that's cool.

Could you try simplify the patch.
I'd say the cache should be move to the `slint-macros`  crate.

Also check if the tests are actually testing something interesting.


~~One corner case i can imaging is two `slint!(export {Foo} from "foo.slint";)` macro, each of them in a different module.  One of them should load `module_a/foo.slint`,  and the other should load `module_b/foo.slint`~~  (edit: not a problem as the path is part of the key

**Comment by tilladam** [comment]
Thanks! Done:

- Moved the cache back into `slint-macros`. For context: it actually started in `slint-macros` — I'd moved it into `i-slint-compiler` only so the benchmark could call the real `key_material` instead of a copy. Now that the benchmark is dropped, there's no reason for it to live there, so it's back in `slint-macros`.
- Simplified the patch: dropped the benchmark module, so it's now just `expansion_cache.rs` plus the macro wiring (+343/−15), all in `slint-macros`.
- Reworked the tests to exercise the real `store`/`lookup` path: a stored key is served back and the hit is asserted via a test-only hit counter, with dependency-change invalidation checked through the public API, plus key replacement and FIFO eviction. Removed the weak "two identical modules compile and agree" test.

And yes — the source path is part of the key, so the two-modules / different-`foo.slint` case stays distinct.

File: api/rs/macros/expansion_cache.rs [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+
+    // Environment / mode that selects the generator or feeds code generation. The
+    // rust-analyzer flag picks the generator, so it must distinguish the entries.
+    let _ = write!(key, "ra={};", is_rust_analyzer());
+    for var in [
+        "SLINT_LIVE_PREVIEW",
+        "SLINT_STYLE",
+        "SLINT_ASSET_SECTION",
+        "CARGO_MANIFEST_DIR",
+        "CARGO_PKG_NAME",
```
Comment (ogoffart) [inline api/rs/macros/expansion_cache.rs]: I don't think environment variable can change while the process is running.

Filtering on the CompilerConfiguration settings should actually be enough.

**Review by ogoffart — APPROVED** [review]
Great.  Looks good to me.

I think the comments could be a bit simplified.  
I tend to prefer concise comment that do not document implementation details too much.

**Review by ogoffart — APPROVED** [review]
